src/sentry/mail/notifications.py

```diff
@@ -21,48 +21,48 @@ def get_headers(notification: BaseNotification) -> Mapping[str, Any]:
         "X-SMTPAPI": json.dumps({"category": notification.get_category()}),
     }
 
-    if notification.group:
+    group = getattr(notification, "group", None)
+    if group:
         headers.update(
             {
-                "X-Sentry-Logger": notification.group.logger,
-                "X-Sentry-Logger-Level": notification.group.get_level_display(),
-                "X-Sentry-Reply-To": group_id_to_email(notification.group.id),
+                "X-Sentry-Logger": group.logger,
+                "X-Sentry-Logger-Level": group.get_level_display(),
+                "X-Sentry-Reply-To": group_id_to_email(group.id),
             }
         )
 
     return headers
 
 
 def get_subject_with_prefix(
-    notification: BaseNotification, mail_option_key: Optional[str] = None
+    notification: BaseNotification,
+    context: Optional[Mapping[str, Any]] = None,
+    mail_option_key: Optional[str] = None,
 ) -> bytes:
     key = mail_option_key or "mail:subject_prefix"
     prefix = str(
         ProjectOption.objects.get_value(notification.project, key)
         or options.get("mail.subject-prefix")
     )
-    return f"{prefix}{notification.get_subject()}".encode("utf-8")
+    return f"{prefix}{notification.get_subject(context)}".encode("utf-8")
 
 
-def get_unsubscribe_link(user_id: int, group_id: int) -> str:
+def get_unsubscribe_link(user_id: int, resource_id: int, key: str = "issue") -> str:
     return generate_signed_link(
         user_id,
-        "sentry-account-email-unsubscribe-issue",
-        kwargs={"issue_id": group_id},
+        f"sentry-account-email-unsubscribe-{key}",
+        kwargs={f"{key}_id": resource_id},
     )
 
 
-def can_users_unsubscribe(notification: BaseNotification) -> bool:
-    return bool(notification.group)
-
-
 def log_message(notification: BaseNotification, user: User) -> None:
     extra = {
         "project_id": notification.project.id,
         "user_id": user.id,
     }
-    if notification.group:
-        extra.update({"group": notification.group.id})
+    group = getattr(notification, "group", None)
```

Comment:
this also seems like maybe it was from the typing PR?

Reply:
I reorganized the classes so that only some of them have the `group` attribute. For example, the deploy notification.